Fix inductor <> ddp_optimizer issue

bdhirsh opened the pull request:
cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @anijain2305 @shunting314 this is a tentative fix for https://github.com/pytorch/pytorch/issues/107362. More discussion at https://github.com/pytorch/pytorch/issues/107362#issuecomment-1696155330.

@wconstab pointed out that inductor found a graph with 6 input mutations and only 1 output, and seemed to be (incorrectly) chopping off the first "6" outputs from the graph (even though there is only 1). It looks like this is because:

(1) AOTAutograd has special handling for input mutations in inference vs. training graphs. In a training graph, whenever AOTAutograd sees an input mutation, it will add an **extra** output to the graph, corresponding to the updated input (and then at runtime, it will grab the updated input, and perform the actual mutation outside of the graph).

In inference, AOTAutograd is smarter and can leave the input mutations directly in the graph for inductor to optimize (doing this in training is harder). In inference, AOTAutograd will **not** add any extra graph outputs for input mutations.

It looks like inductor was unconditionally assuming that input mutations counted as extra outputs in the graph, which is wrong for the inference case.


Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #108570
* #108243
* #108235
* __->__ #108081



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

bdhirsh committed:
Fix inductor <> ddp_optimizer issue

[ghstack-poisoned]

wconstab reviewed (approved):
thanks for PR'ing this fix! would be good to add a test, i can turn the orig PR repro into one in a follow up.

bdhirsh commented:
Sounds good - let me try to make a simpler tests (should be doable just by using inductor + inference graph + input mutations + layout optimization)

bdhirsh commented:
Oh I'll also add an assertion that we don't go beyond the number of graph outputs (which  would also have caught this issue instead of silently not preserving strides for certain outputs)

bdhirsh commented:
Hmm I tried creating a repro by taking the example from the linked issue, removing the DDP code, and manually inserting a graph break at the same place that the ddp optimizer would. I confirmed that between the original repro and my modified version the AOTAutograd graphs that we send to inductor are identical, but for some reason I can't trigger the issue.

I'll try inspecting the generating inductor code and see if it's any different

shunting314 commented:
@bdhirsh that maybe expected since the graph break caused by DDPOptimizer has subtile difference compared to other graph breaks in dynamo. Check https://github.com/pytorch/pytorch/issues/102591 for a bit more details.

bdhirsh commented:
hmm @shunting314 - will pointed out to me that the issue you linked only affects code that's run under `TORCHINDUCTOR_KEEP_OUTPUT_STRIDE=0`. But my repro isn't using that env var - so that issue is probably unrelated?

shunting314 commented:
@bdhirsh so my point is that for the manually injected graph breaks, downstream graph already get the output of ***OPTIMIZED*** upstream graph, whether we succeed to decide the correct output tensors to keep their strides does not matter. Even if we the compiler change the output tensor's stride, the downstream graph will get tensors with those updated stride. Just want to explain why you could not repro earlier and I think we have to use DDPOptimizer to repro the issue.

wconstab commented:
i'm not sure how brian did the repro, but i would expect you could repro without DDPOptimizer and without graph-breaks, just by 
1) creating a model with some input mutations
2) running its forward inside no_grad, so inductor incorrectly ignores preserving output strides
3) asserting that the output tensors from the model have the same strides as eager had when running the model

Why is DDPoptimizer needed for the repro?

I do understand that if graph-breaks are introduced by dynamo, dynamo works around compilation, ensuring it will match expectations.

bdhirsh committed:
Update on "Fix inductor <> ddp_optimizer issue"

cc shunting314 this is a tentative fix for https://github.com/pytorch/pytorch/issues/107362. More discussion at https://github.com/pytorch/pytorch/issues/107362#issuecomment-1696155330.

wconstab pointed out that inductor found a graph with 6 input mutations and only 1 output, and seemed to be (incorrectly) chopping off the first "6" outputs from the graph (even though there is only 1). It looks like this is because:

(1) AOTAutograd has special handling for input mutations in inference vs. training graphs. In a training graph, whenever AOTAutograd sees an input mutation, it will add an **extra** output to the graph, corresponding to the updated input (and then at runtime, it will grab the updated input, and perform the actual mutation outside of the graph).

In inference, AOTAutograd is smarter and can leave the input mutations directly in the graph for inductor to optimize (doing this in training is harder). In inference, AOTAutograd will **not** add any extra graph outputs for input mutations.

It looks like inductor was unconditionally assuming that input mutations counted as extra outputs in the graph, which is wrong for the inference case.





cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

bdhirsh commented on test/dynamo/test_repros.py:
```diff
@@ -3270,6 +3270,29 @@ def f():
 
         self.assertEqual(f(), _GLOBAL_CPU_TENSOR + _GLOBAL_CPU_TENSOR)
 
+    @torch._inductor.config.patch("layout_optimization", True)
```
@wconstab got a repro! It just took a bit more work than I thought in order to get inductor to actually change the output stride of the convolution. I needed:
- layout_optimization = True
- channels dim to be > 64 (from the inductor code [here](https://github.com/pytorch/pytorch/blob/main/torch/_inductor/graph.py#L325))

wconstab replied:
great, thanks @bdhirsh!!

bdhirsh committed:
Update on "Fix inductor <> ddp_optimizer issue"

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov anijain2305 shunting314 this is a tentative fix for https://github.com/pytorch/pytorch/issues/107362. More discussion at https://github.com/pytorch/pytorch/issues/107362#issuecomment-1696155330.

wconstab pointed out that inductor found a graph with 6 input mutations and only 1 output, and seemed to be (incorrectly) chopping off the first "6" outputs from the graph (even though there is only 1). It looks like this is because:

(1) AOTAutograd has special handling for input mutations in inference vs. training graphs. In a training graph, whenever AOTAutograd sees an input mutation, it will add an **extra** output to the graph, corresponding to the updated input (and then at runtime, it will grab the updated input, and perform the actual mutation outside of the graph).

In inference, AOTAutograd is smarter and can leave the input mutations directly in the graph for inductor to optimize (doing this in training is harder). In inference, AOTAutograd will **not** add any extra graph outputs for input mutations.

It looks like inductor was unconditionally assuming that input mutations counted as extra outputs in the graph, which is wrong for the inference case.





cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

bdhirsh committed:
Update on "Fix inductor <> ddp_optimizer issue"

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov anijain2305 shunting314 this is a tentative fix for https://github.com/pytorch/pytorch/issues/107362. More discussion at https://github.com/pytorch/pytorch/issues/107362#issuecomment-1696155330.

wconstab pointed out that inductor found a graph with 6 input mutations and only 1 output, and seemed to be (incorrectly) chopping off the first "6" outputs from the graph (even though there is only 1). It looks like this is because:

(1) AOTAutograd has special handling for input mutations in inference vs. training graphs. In a training graph, whenever AOTAutograd sees an input mutation, it will add an **extra** output to the graph, corresponding to the updated input (and then at runtime, it will grab the updated input, and perform the actual mutation outside of the graph).

In inference, AOTAutograd is smarter and can leave the input mutations directly in the graph for inductor to optimize (doing this in training is harder). In inference, AOTAutograd will **not** add any extra graph outputs for input mutations.

It looks like inductor was unconditionally assuming that input mutations counted as extra outputs in the graph, which is wrong for the inference case.





cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]
